### PR TITLE
Space - Stephanie

### PR DESCRIPTION
**Pull request by stephaniejmars**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I made a top ten model and a spotlight model method
Describe how you approached testing that model method. What edge cases did you come up with? | I approached what was in the lessons and built from there. 
What are `session` and `flash`? What is the difference between them? | Session keeps track of the user thats logged in and flash is a message sent to the user. 
What was one thing that you gained more clarity on through this assignment? | Relationships and a better understanding of all the moving parts. 
What is the Heroku URL of your deployed application? | https://stephanies-media-ranker.herokuapp.com/

# Assignment Submission: Media Ranker

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+Rails.application.routes.draw do
+  root to: 'homepages#index'
+
+  resources :works
```
Comment (dHelmgren) [inline config/routes.rb:4]: Weird that your edit page for works isn't working. I'm going to take a peek at that page too!

File: app/views/homepages/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    <section class="top-ten__list-container">
+      <h2 class="top-ten__header">Top Books</h2>
+        <ul class="list-group top-ten__list">
+          <%@top_ten_book.each do |book|%>
+            <li class="list-group-item">
+            <h4><%=link_to book.title, work_path(book.id)%>
+            <small class="top-ten__creator">by: <%=book.creator%></small>
+            </h4> 
+            <p><%=book.votes.count%> votes</p>
+            </li>
```
Comment (dHelmgren) [inline app/views/homepages/index.html.erb:32]: This is a great place to use a view partial, as you have the same html repeated 3 times!!!

File: app/views/works/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,41 @@
+<section>
+  <h2><%= @work.title %> </h2>
+  <p>Created by: <%= @work.creator %> </p>
+  <p>Published: <%= @work.publication_year %> </p>
+  <p><%= @work.description %></p>
+</section>
+<section>
+    <span><%= button_to "Back to media ranks", root_path, :class => "btn btn-primary", form: {style: 'display:inline-block;'}%></span>
+    <span><%= button_to "Edit", edit_work_path, :class => "btn btn-primary", form: {style: 'display:inline-block;'}%></span>
```
Comment (dHelmgren) [inline app/views/works/show.html.erb:9]: I think both of these buttons are trying to use the `post` verb, which is breaking the routes!

File: app/controllers/works_controller.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    if @work.nil?
+      head :not_found
+      return
+    end
+  end
+
+  def edit
+    @work = Work.find_by(id: params[:id])
+
+    if @work.nil?
```
Comment (dHelmgren) [inline app/controllers/works_controller.rb:38]: You have a few methods that require a specific work to be fetched, making this a good place to use controller filters.

File: app/controllers/votes_controller.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  def create
+    if !session[:user_id]
+      flash[:message] = "A problem occured: You must log in to do that"
+      redirect_back(fallback_location: root_path)
+      return
+    elsif Vote.find_by(user_id: session[:user_id], work_id: params[:work_id])
+      flash[:message] = "You have already voted for this work"
+      redirect_back(fallback_location: root_path)
+      return
+    elsif session[:user_id]
```
Comment (dHelmgren) [inline app/controllers/votes_controller.rb:11]: This doesn't need to be an `elsif`, as you know after the initial `if` that we at least have a user.

**Comment by dHelmgren** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | route is missing
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | not sorted by vote
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | no 404
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate |
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user |
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work |
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works |
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | no relationship tests
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | blank relationship tests
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | no relationship tests
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | no business logic tests

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review |
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention |
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention | ✔️

### Additional Feedback

It feels to me like the tail end of this assignment was rushed. I see a lot of really solid foundational work, but the newer concepts didn't make it in to the project. I really want to highlight testing in all of this, because if you had tested this project fully, that on its own would have made the difference between a green and a red assignment. Message me about this project if you want to talk through my feedback!

This is one assignment. Don't let this stumbling block bring you down!
